File: main/contrib/Windows API Code Pack 1.1/source/Samples/DirectX/VB/UtilitiesLibrary/D3DUtilities/Math/Matrix.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.


Imports Microsoft.VisualBasic
Imports System
Imports System.Runtime.InteropServices
Imports Microsoft.WindowsAPICodePack.DirectX.Direct3D

Namespace Microsoft.WindowsAPICodePack.DirectX.DirectXUtilities
	<StructLayout(LayoutKind.Sequential)> _
	Public Class MatrixMath
		#Region "operator *"

		Public Shared Function VectorMultiply(ByVal a As Matrix4x4F, ByVal b As Vector4F) As Vector4F
			Return New Vector4F(a.M11 * b.x + a.M12 * b.y + a.M13 * b.z + a.M14 * b.w, a.M21 * b.x + a.M22 * b.y + a.M23 * b.z + a.M24 * b.w, a.M31 * b.x + a.M32 * b.y + a.M33 * b.z + a.M34 * b.w, a.M41 * b.x + a.M42 * b.y + a.M43 * b.z + a.M44 * b.w)
		End Function
		#End Region

		#Region "MatrixScale"
		Public Shared Function MatrixScale(ByVal x As Single, ByVal y As Single, ByVal z As Single) As Matrix4x4F
			Return New Matrix4x4F(x, 0, 0, 0, 0, y, 0, 0, 0, 0, z, 0, 0, 0, 0, 1)
		End Function
		#End Region

		#Region "MatrixTranslate"
		Public Shared Function MatrixTranslate(ByVal x As Single, ByVal y As Single, ByVal z As Single) As Matrix4x4F
			Return New Matrix4x4F(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, x, y, z, 1)
		End Function
		#End Region

		#Region "MatrixRotationX"
		Public Shared Function MatrixRotationX(ByVal angle As Single) As Matrix4x4F
			Dim sin As Single = CSng(Math.Sin(angle))
			Dim cos As Single = CSng(Math.Cos(angle))
			Return New Matrix4x4F(1, 0, 0, 0, 0, cos, sin, 0, 0, -sin, cos, 0, 0, 0, 0, 1)
		End Function
		#End Region

		#Region "MatrixRotationY"
		Public Shared Function MatrixRotationY(ByVal angle As Single) As Matrix4x4F
			Dim sin As Single = CSng(Math.Sin(angle))
			Dim cos As Single = CSng(Math.Cos(angle))
			Return New Matrix4x4F(cos, 0, -sin, 0, 0, 1, 0, 0, sin, 0, cos, 0, 0, 0, 0, 1)
		End Function
		#End Region

		#Region "MatrixRotationZ"
		Public Shared Function MatrixRotationZ(ByVal angle As Single) As Matrix4x4F
			Dim sin As Single = CSng(Math.Sin(angle))
			Dim cos As Single = CSng(Math.Cos(angle))
			Return New Matrix4x4F(cos, sin, 0, 0, -sin, cos, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1)
		End Function
		#End Region
	End Class

End Namespace

File: main/contrib/Windows API Code Pack 1.1/source/Samples/DirectX/CS/UtilitiesLibrary/D3DUtilities/Math/Matrix.cs
// Copyright (c) Microsoft Corporation.  All rights reserved.

using System;
using System.Runtime.InteropServices;
using Microsoft.WindowsAPICodePack.DirectX.Direct3D;

namespace Microsoft.WindowsAPICodePack.DirectX.DirectXUtilities
{
    [StructLayout(LayoutKind.Sequential)]
    public class MatrixMath
    {
        #region operator *

        public static Vector4F VectorMultiply(Matrix4x4F a, Vector4F b)
        {
            return new Vector4F(
                a.M11 * b.X + a.M21 * b.Y + a.M31 * b.Z + a.M41 * b.W,
                a.M12 * b.X + a.M22 * b.Y + a.M32 * b.Z + a.M42 * b.W,
                a.M13 * b.X + a.M23 * b.Y + a.M33 * b.Z + a.M43 * b.W,
                a.M14 * b.X + a.M24 * b.Y + a.M34 * b.Z + a.M44 * b.W
                );
        }
        #endregion

        #region MatrixScale
        public static Matrix4x4F MatrixScale(float x, float y, float z)
        {
            return new Matrix4x4F(
                x, 0, 0, 0,
                0, y, 0, 0,
                0, 0, z, 0,
                0, 0, 0, 1
                );
        } 
        #endregion

        #region MatrixTranslate
        public static Matrix4x4F MatrixTranslate(float x, float y, float z)
        {
            return new Matrix4x4F(
                1, 0, 0, 0,
                0, 1, 0, 0,
                0, 0, 1, 0,
                x, y, z, 1
                );
        }
        #endregion

        #region MatrixRotationX
        public static Matrix4x4F MatrixRotationX(float angle)
        {
            float sin = (float)Math.Sin(angle);
            float cos = (float)Math.Cos(angle);
            return new Matrix4x4F(
                1, 0, 0, 0,
                0, cos, sin, 0,
                0, -sin, cos, 0,
                0, 0, 0, 1
                );
        }
        #endregion

        #region MatrixRotationY
        public static Matrix4x4F MatrixRotationY(float angle)
        {
            float sin = (float)Math.Sin(angle);
            float cos = (float)Math.Cos(angle);
            return new Matrix4x4F(
                cos, 0, -sin, 0,
                0, 1, 0, 0,
                sin, 0, cos, 0,
                0, 0, 0, 1
                );
        }
        #endregion

        #region MatrixRotationZ
        public static Matrix4x4F MatrixRotationZ(float angle)
        {
            float sin = (float)Math.Sin(angle);
            float cos = (float)Math.Cos(angle);
            return new Matrix4x4F(
                cos, sin, 0, 0,
                -sin, cos, 0, 0,
                0, 0, 1, 0,
                0, 0, 0, 1
                );
        }
        #endregion
    }

}
